zk/da/blob_da.go

```diff
@@ -0,0 +1,108 @@
+package da
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/erigontech/erigon-lib/common"
+	"github.com/erigontech/erigon-lib/common/hexutil"
+	"github.com/erigontech/erigon/zk/hermez_db"
+	"github.com/erigontech/erigon/zk/tx"
+	"github.com/erigontech/erigon/zkevm/jsonrpc/client"
+)
+
+type OffChainBlob struct {
+	TxHash      string      `json:"txHash"`
+	BlobInputs  []BlobInput `json:"blobInputs"`
+	StartBatch  uint64      `json:"startBatch"`
+	EndBatch    uint64      `json:"endBatch"`
+	Compression uint64      `json:"compression"`
+	Timestamp   int64       `json:"timestamp"`
+}
+
+type BlobInput struct {
+	BatchNumber    string         `json:"batchNumber"`
+	Coinbase       common.Address `json:"coinbase"`
+	LimitTimestamp string         `json:"limitTimestamp"`
+	BatchL2Data    string         `json:"batchL2Data"`
+}
+
+func GetOffChainBlobs(url string, limit, offset uint64) ([]*OffChainBlob, bool, error) {
+	res, err := client.JSONRPCCall(url, "data_getOffChainBlobs", limit, offset)
+	if err != nil {
+		return nil, false, err
+	}
+
+	var ocb []*OffChainBlob
+	if err = json.Unmarshal(res.Result, &ocb); err != nil {
+		return nil, false, err
+	}
+
+	// we have reached the end of the blobs
+	if ocb == nil {
+		return nil, true, nil
+	}
+
+	return ocb, false, nil
+}
+
+func CreateL1BatchDataFromBlobInput(hermezDb *hermez_db.HermezDb, input BlobInput) (uint64, []byte, error) {
+	batchNumber, err := hexutil.DecodeUint64(input.BatchNumber)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	batchData, err := hexutil.Decode(input.BatchL2Data)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	limitTimestamp, err := hexutil.Decode(input.LimitTimestamp)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	allForks, allBatches, err := hermezDb.GetAllForkHistory()
+	if err != nil {
+		return 0, nil, err
+	}
+
+	var forkId uint64
+	for idx, batch := range allBatches {
+		if batchNumber >= batch {
+			forkId = allForks[idx]
+		}
+	}
+
+	decodedBatchL2Data, err := tx.DecodeBatchL2Blocks(batchData, forkId)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	var highestIndexUsed uint32
+	for _, block := range decodedBatchL2Data {
+		if block.L1InfoTreeIndex > highestIndexUsed {
+			highestIndexUsed = block.L1InfoTreeIndex
+		}
+	}
+
+	l1InfoRoot, err := hermezDb.GetL1InfoRootByIndex(uint64(highestIndexUsed))
```

Comment:
We have a problem here, the l1 info root should be a high water mark across all batches as we discovered looking at the L1 the other day.  This mechanism will only report the highest info root used within a single batch.  I think the caller of this function within the loop in the stage itself should pass in the latest returned info root, then if we find the highest used index is 0 in this batch then we can just use the last info root for this batch.  I hope that makes sense?

Reply:
To go a little more detailed here, the caller should hold on to the last returned L1 info root as a new return from this function.  Then pass it back into the next call to this function so we can always use the latest value in case of no new indexes used in a batch.

The potential issue I see here is that we use the info root whilst verifying a bad batch during recovery, in theory if there are no info tree indexes used in a batch we should be OK, but better to keep in line with how the L1 data handles these info roots.